graphql_compiler/schema/__init__.py

```diff
@@ -340,36 +340,47 @@ def _serialize_date(value: Any) -> str:
 
 def _parse_date_value(value: Any) -> date:
     """Deserialize a Date object from its proper ISO-8601 representation."""
-    return arrow.get(value, "YYYY-MM-DD").date()
+    if type(value) == date:
+        # We prefer exact type equality instead of isinstance() because datetime objects
+        # are subclasses of date but are not interchangeable for dates for our purposes.
+        return value
+    elif isinstance(value, str):
+        return arrow.get(value, "YYYY-MM-DD").date()
```

Comment:
It seems incongruous to me that we disallow subtypes of `date` to prevent implicit loss of precision, but then implicitly drop precision ourselves when presented with string-specified date-times.

```python
>>> arrow.get('0001-02-03 04:05:06', "YYYY-MM-DD").date()
datetime.date(1, 2, 3)
```

I think there are reasonable arguments to be made in favor of either behavior, and I'd rather we choose one and stick with it.  (If we decide to stick with disallowing implicit loss of precision, then we should start emitting warnings for the string case now and switch to raising exceptions in the 2.0 release.)

Reply:
This is a good point. I think we should not allow implicit loss of precision, and we're pretty committed to only releasing 2.0-series versions at this point (the `main` branch already has breaking changes), so I'd like to go straight to raising exceptions for loss of precision in the interest of time and simplicity.

Reply:
Sounds good!  (I agree on disallowing implicit loss of precision.)